psi4/src/psi4/libsapt_solver/usapt0.cc

```diff
@@ -0,0 +1,2535 @@
+
+    start = time(NULL);
+
+    outfile->Printf( "    -----------------------------------------\n");
+    outfile->Printf( "    %-4s %11s  %11s  %10s\n", "Iter", "Monomer A", "Monomer B", "Time [s]");
+    outfile->Printf( "    -----------------------------------------\n");
+
+    int iter;
+    for (iter = 0; iter < maxiter_; iter++) {
+
+        std::map<std::string, std::shared_ptr<Matrix> > b;
+        if (r2A > delta_) {
+            b["Aa"] = pa_A;
+            b["Ab"] = pb_A;
+        }
+        if (r2B > delta_) {
+            b["Ba"] = pa_B;
+            b["Bb"] = pb_B;
+        }
+
+        std::map<std::string, std::shared_ptr<Matrix> > s =
+            product(b);
+
+        if (r2A > delta_) {
+            std::shared_ptr<Matrix> sa_A = s["Aa"];
+            std::shared_ptr<Matrix> sb_A = s["Ab"];
+            double alpha = (ra_A->vector_dot(za_A) + rb_A->vector_dot(zb_A))
+                           / (pa_A->vector_dot(sa_A) + pb_A->vector_dot(sb_A));
+            if (alpha < 0.0) {
+                throw PSIEXCEPTION("Monomer A: A Matrix is not SPD");
+            }
+            int no = xa_A_->nrow();
+            int nv = xa_A_->ncol();
+            double** xp = xa_A_->pointer();
+            double** rp = ra_A->pointer();
+            double** pp = pa_A->pointer();
+            double** sp = sa_A->pointer();
+            C_DAXPY(no*nv, alpha,pp[0],1,xp[0],1);
+            C_DAXPY(no*nv,-alpha,sp[0],1,rp[0],1);
+            r2A = C_DDOT(no*nv,rp[0],1,rp[0],1);
+            no = xb_A_->nrow();
+            nv = xb_A_->ncol();
+            xp = xb_A_->pointer();
+            rp = rb_A->pointer();
+            pp = pb_A->pointer();
+            sp = sb_A->pointer();
+            C_DAXPY(no*nv, alpha,pp[0],1,xp[0],1);
+            C_DAXPY(no*nv,-alpha,sp[0],1,rp[0],1);
+            r2A += C_DDOT(no*nv,rp[0],1,rp[0],1);
+            r2A = sqrt(r2A) / b2A;
+        }
+
+        if (r2B > delta_) {
+            std::shared_ptr<Matrix> sa_B = s["Ba"];
+            std::shared_ptr<Matrix> sb_B = s["Bb"];
+            double alpha = (ra_B->vector_dot(za_B) + rb_B->vector_dot(zb_B))
+                           / (pa_B->vector_dot(sa_B) + pb_B->vector_dot(sb_B));
+            if (alpha < 0.0) {
+                throw PSIEXCEPTION("Monomer B: A Matrix is not SPD");
+            }
+            int no = xa_B_->nrow();
+            int nv = xa_B_->ncol();
+            double** xp = xa_B_->pointer();
+            double** rp = ra_B->pointer();
+            double** pp = pa_B->pointer();
+            double** sp = sa_B->pointer();
+            C_DAXPY(no*nv, alpha,pp[0],1,xp[0],1);
+            C_DAXPY(no*nv,-alpha,sp[0],1,rp[0],1);
+            r2B = C_DDOT(no*nv,rp[0],1,rp[0],1);
+            no = xb_B_->nrow();
+            nv = xb_B_->ncol();
+            xp = xb_B_->pointer();
+            rp = rb_B->pointer();
+            pp = pb_B->pointer();
+            sp = sb_B->pointer();
+            C_DAXPY(no*nv, alpha,pp[0],1,xp[0],1);
```

Comment:
Matrix objects have an AXPY function now so you could replace a lot of this with `xa_B_->axpy(alpha, pa_B)`. The DDOT call below could be replace with a `ra_B->sum_of_squares();` as well. Could help with readability in the future.